File: database/js/repositories/sanctionRepo.js
// Repo Sanction - opérations CRUD

const { dbGet, dbRun, dbAll } = require('../index');
const Sanction = require('../models/Sanction');

function toSQLiteDateValue(value) {
  if (!value) return null;
  if (value instanceof Date) return value.toISOString();
  return String(value);
}

const sanctionRepo = {
  async create(sanction) {
    const s = sanction instanceof Sanction ? sanction : new Sanction(sanction);
    const data = s.toDatabase();

    const providedExpiresAt = toSQLiteDateValue(data.expires_at);
    const duration = data.duration ?? null; // secondes

    // Si expires_at n'est pas fourni mais duration oui, calcul auto côté SQLite.
    const res = await dbRun(
      `
      INSERT INTO sanctions
        (guild_id, user_id, moderator_id, type, reason, duration, expires_at, active, created_at)
      VALUES
        (?, ?, ?, ?, ?, ?,
          COALESCE(
            ?,
            CASE
              WHEN ? IS NOT NULL THEN datetime(CURRENT_TIMESTAMP, '+' || ? || ' seconds')
              ELSE NULL
            END
          ),
          ?,
          CURRENT_TIMESTAMP
        )
      `,
      [
        data.guild_id,
        data.user_id,
        data.moderator_id,
        data.type,
        data.reason,
        duration,
        providedExpiresAt,
        duration,
        duration,
        data.active ? 1 : 0,
      ]
    );

    return this.findById(res.lastID);
  },

  async findById(id) {
    const row = await dbGet('SELECT * FROM sanctions WHERE id = ?', [id]);
    return row ? new Sanction(row) : null;
  },

  async findByUser(userId, guildId) {
    const rows = await dbAll(
      `
      SELECT * FROM sanctions
      WHERE user_id = ? AND guild_id = ?
      ORDER BY datetime(created_at) DESC
      `,
      [userId, guildId]
    );
    return rows.map((r) => new Sanction(r));
  },

  async findActiveByUser(userId, guildId) {
    const rows = await dbAll(
      `
      SELECT * FROM sanctions
      WHERE user_id = ? AND guild_id = ?
        AND active = 1
        AND (expires_at IS NULL OR datetime(expires_at) > CURRENT_TIMESTAMP)
      ORDER BY datetime(created_at) DESC
      `,
      [userId, guildId]
    );
    return rows.map((r) => new Sanction(r));
  },

  async findActiveByType(guildId, type) {
    const rows = await dbAll(
      `
      SELECT * FROM sanctions
      WHERE guild_id = ?
        AND type = ?
        AND active = 1
        AND (expires_at IS NULL OR datetime(expires_at) > CURRENT_TIMESTAMP)
      ORDER BY datetime(created_at) DESC
      `,
      [guildId, type]
    );
    return rows.map((r) => new Sanction(r));
  },

  async findExpired() {
    const rows = await dbAll(
      `
      SELECT * FROM sanctions
      WHERE active = 1
        AND expires_at IS NOT NULL
        AND datetime(expires_at) <= CURRENT_TIMESTAMP
      ORDER BY datetime(expires_at) ASC
      `
    );
    return rows.map((r) => new Sanction(r));
  },

  async deactivate(id) {
    await dbRun('UPDATE sanctions SET active = 0 WHERE id = ?', [id]);
    return this.findById(id);
  },

  async deactivateByUserAndType(userId, guildId, type) {
    const res = await dbRun(
      `
      UPDATE sanctions
      SET active = 0
      WHERE user_id = ? AND guild_id = ? AND type = ? AND active = 1
      `,
      [userId, guildId, type]
    );
    return res.changes;
  },

  async getRecent(guildId, limit = 50) {
    const rows = await dbAll(
      `
      SELECT * FROM sanctions
      WHERE guild_id = ?
      ORDER BY datetime(created_at) DESC
      LIMIT ?
      `,
      [guildId, limit]
    );
    return rows.map((r) => new Sanction(r));
  },

  async getStats(guildId) {
    return dbAll(
      `
      SELECT
        type,
        COUNT(*) AS total,
        SUM(CASE WHEN active = 1 THEN 1 ELSE 0 END) AS active
      FROM sanctions
      WHERE guild_id = ?
      GROUP BY type
      `,
      [guildId]
    );
  },

  /**
   * Révoque tous les mutes actifs pour un utilisateur dans une guild
   * et enregistre une sanction "unmute" pour l'historique.
   */
  async revokeActiveMute(guildId, userId, moderatorId, reason) {
    const revokedCount = await this.deactivateByUserAndType(userId, guildId, 'mute');

    const unmuteSanction = await this.create({
      guildId,
      userId,
      moderatorId,
      type: 'unmute',
      reason: reason || 'Unmute sans raison spécifiée',
      duration: null,
    });

    return { revokedCount, unmuteSanction };
  },
};

module.exports = sanctionRepo;

File: database/python/repositories/sanction_repo.py
# Repo Sanction Python
from database.python.connection import get_connection, dict_from_row

def get_by_guild(guild_id: str, limit: int = 100):
    """Toutes les sanctions d'une guild."""
    with get_connection() as conn:
        cursor = conn.execute("""
            SELECT s.*, u.username
            FROM sanctions s
            LEFT JOIN users u ON s.user_id = u.discord_id AND s.guild_id = u.guild_id
            WHERE s.guild_id = ?
            ORDER BY s.created_at DESC
            LIMIT ?
        """, (guild_id, limit))
        return [dict_from_row(row) for row in cursor.fetchall()]

def get_by_user(user_id: str, guild_id: str):
    """Sanctions d'un user."""
    with get_connection() as conn:
        cursor = conn.execute("""
            SELECT s.*, u.username
            FROM sanctions s
            LEFT JOIN users u ON s.user_id = u.discord_id AND s.guild_id = u.guild_id
            WHERE s.user_id = ? AND s.guild_id = ?
            ORDER BY s.created_at DESC
        """, (user_id, guild_id))
        return [dict_from_row(row) for row in cursor.fetchall()]

def get_active(guild_id: str):
    """Sanctions actives (bans/mutes en cours)."""
    with get_connection() as conn:
        cursor = conn.execute("""
            SELECT s.*, u.username
            FROM sanctions s
            LEFT JOIN users u ON s.user_id = u.discord_id AND s.guild_id = u.guild_id
            WHERE s.guild_id = ? AND s.active = 1
            ORDER BY s.created_at DESC
        """, (guild_id,))
        return [dict_from_row(row) for row in cursor.fetchall()]

def get_by_type(guild_id: str, sanction_type: str):
    """Sanctions par type (ban, mute, kick)."""
    with get_connection() as conn:
        cursor = conn.execute("""
            SELECT s.*, u.username
            FROM sanctions s
            LEFT JOIN users u ON s.user_id = u.discord_id AND s.guild_id = u.guild_id
            WHERE s.guild_id = ? AND s.type = ?
            ORDER BY s.created_at DESC
            LIMIT 100
        """, (guild_id, sanction_type))
        return [dict_from_row(row) for row in cursor.fetchall()]

def get_stats_by_type(guild_id: str):
    """Stats par type pour pie chart."""
    with get_connection() as conn:
        cursor = conn.execute("""
            SELECT type, COUNT(*) as count
            FROM sanctions
            WHERE guild_id = ?
            GROUP BY type
        """, (guild_id,))
        return [dict_from_row(row) for row in cursor.fetchall()]

def get_stats_by_day(guild_id: str, days: int = 30):
    """Sanctions par jour pour line chart."""
    with get_connection() as conn:
        cursor = conn.execute("""
            SELECT DATE(created_at) as date, type, COUNT(*) as count
            FROM sanctions
            WHERE guild_id = ?
            AND created_at >= datetime('now', ?)
            GROUP BY DATE(created_at), type
            ORDER BY date
        """, (guild_id, f'-{days} days'))
        return [dict_from_row(row) for row in cursor.fetchall()]

def create(sanction_data: dict):
    """Créer une nouvelle sanction."""
    with get_connection() as conn:
        cursor = conn.execute("""
            INSERT INTO sanctions (guild_id, user_id, moderator_id, type, reason, duration, expires_at, active)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        """, (
            sanction_data.get('guild_id'),
            sanction_data.get('user_id'),
            sanction_data.get('moderator_id'),
            sanction_data.get('type'),
            sanction_data.get('duration'),
            sanction_data.get('expires_at'),
            sanction_data.get('active', 1)
        ))
        conn.commit()
        return cursor.lastrowid

def deactivate(sanction_id: int):
    """Désactiver une sanction."""
    with get_connection() as conn:
        conn.execute(
            "UPDATE sanctions SET active = 0 WHERE id = ?",
            (sanction_id,)
        )
        conn.commit()
